src/System.Reflection.MetadataLoadContext/src/System/Reflection/PathAssemblyResolver.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.IO;
+
+namespace System.Reflection
+{
+    /// <summary>
+    /// A MetadataAssemblyResolver that uses paths to each assembly.
+    /// Assemblies cannot be duplicated by simlple name.
+    /// The assembly's simple name is used as a key to obtain the corresponding directory.
+    /// </summary>
+    /// <remarks>
+    /// The assembly name does not need to include the assembly file extension (.dll). If an extension exists
+    /// it is removed.
+    /// </remarks>
+    /// <exception cref="System.ArgumentException">Thrown when the assembly simple name already exists.</exception>
```

Comment:
This should be on the constructor, not the class.

Also, it should say that the assemblyPaths must not have two files with the same simple name.

Finally, I'd prefer if the exception was thrown explicitly instead of letting dictionary throw, which would look like a bug in the call stack. And assemblyPaths should probably be the paramName.

(I am still nervous about this restriction. It will break valid (not common, but still valid) projects to insist that all assemblies on @(ReferencePath) have a unique name so I'm not sure this API addresses the build scenario correctly and I thought that was the purpose of it. I'm almost tempted to say let's not have it if we're just going to do the trivial case.)

Reply:
I will clarify the comments.

An explicit throw here will cause a tiny perf hit since it must call Contains - it would be nice if there's a TryAdd() method...

What do you suggest to fix the restriction? If we want to emulate the CoreClr host (to a 99% level) then we need to load each assembly (that has the same simple name) and pick the highest assemblyversion+fileversion, and\or we could also place that code in a protected virtual method `ResolveDuplicate` that could be overridden.

Reply:
>  If we want to emulate the CoreClr host

This API is geared towards compilers/tools/ It should be taking Roslyn as the gold standard, not CoreClr host. What does Roslyn do for duplicates?

Reply:
Yes, this is what I've been saying: 

https://github.com/dotnet/corefx/pull/33201#discussion_r230413714

> a good default would be to study what csc does and mimic it as much as possible. A common use case for this API is to use the same references as csc and it would be desirable for this and csc to bind them the same way.

How about this: we cut PathAssemblyResolver from the initial check-in and open an issue to track adding it. I would prefer shipping v1 without it than picking this policy without exploring csc parity. This policy is trivial to implement yourself so we're not holding much value back. I think it's probably enough work to match csc to treat as a distinct feature/review.

This looks like it might be the Roslyn code:

http://source.roslyn.io/#Microsoft.CodeAnalysis/ReferenceManager/CommonReferenceManager.Resolution.cs,99d93dd36939a602,references

Note that the warning it mentions, is usually disabled, and I've suggested removing it from the compiler:

https://github.com/dotnet/roslyn/issues/19640

Reply:
Adding @jaredpar and @tmat: Is that the right source of truth for how the compiler binds assembly references?

Reply:
I'm OK with removing it for now. Anecdotal evendence shows that the compiler will pick the last assembly reference. And it only bases conflicts on the actual assembly name internally not the physical file name.

However, this is can be more complicated as msbuild also has rules for packages - search for `Error CS1703:"Multiple assemblies with equivalent identity have been imported"`

Reply:
> And it only bases conflicts on the actual assembly name internally not the physical file name.

Then we should do the same. 

> However, this is can be more complicated as msbuild also has rules for packages

We should ignore anything higher than csc in the build for the specification here. The rest of the msbuild targets resolve references down to paths. Once you have a set of paths as here, csc is the gold standard.

>  Error CS1703:"Multiple assemblies with equivalent identity have been imported"

This is for the case where the full strong name is duplicated. We can error for that too.

Reply:
I will cut PathAssemblyResolver for now until we can properly add requirements.